open-craft/opencraft
# [FAL-2031] Add Redis support for Ocim provisioned instances

This Pull Request adds Redis cache support to Ocim. The changes won't take effect on already provisioned servers, but new ones or those that explicitly set to use Redis over RabbitMQ. This way we can slowly roll out changes for every server while we are not generating more work when new instances are created.

**~Dependencies~ Related PRs**:

* https://github.com/celery/kombu/pull/1349
* https://github.com/celery/kombu/pull/1351
* https://github.com/edx/edx-platform/pull/28020

**Screenshots**: 

_Admin site integration of `RedisServer`_

![Screenshot 2021-07-01 at 16 10 00](https://user-images.githubusercontent.com/19173947/124138298-ce818680-da86-11eb-90b4-1acc52aab5dc.png)

_Instance admin configuration_

![Screenshot 2021-07-01 at 14 49 00](https://user-images.githubusercontent.com/19173947/124138200-b3af1200-da86-11eb-84c6-2f203ab94e49.png)

_Redis cluster showing the Ocim created user and its ACL_

<img width="1297" alt="Screenshot 2021-07-01 at 15 35 27" src="https://user-images.githubusercontent.com/19173947/124136488-20291180-da85-11eb-97e5-004a37d170f4.png">

**Sandbox URL**: Upon request, I can create an ngrok URL for my local Ocim

**Merge deadline**: None

**Testing instructions**:

1. Checkout the changes on Ocim stage 
2. Create a Redis user and ACL for Ocim if no user was created (or use the default user) and double-check the connection to the Redis server
3. Set the necessary environment variables in the `.env` file (ie. `DEFAULT_INSTANCE_REDIS_URL` )
4. Restart Ocim
5. Provision a new app server
6. Login with edX user
7. Go to instructor/download data tab
8. Download a grades report (as that uses celery, that would use Redis now)
9. Validate the instructor task is finished successfully

A note about staging env: The instance creation nowadays is hectic -- it seems for some instances the LB URL registration did not work.

**Author notes and concerns**:

N/A

**Reviewers**
